tests/test-dladdr.c

```diff
@@ -0,0 +1,177 @@
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr( void *addr, char *addrsym, ExpectedResult expected_result )
+{
+    Dl_info info;
+    int result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match)
+                 || (expected_result == PassWithDifferentAddress && sym_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+    }
+    printf( "check address %p which has symbol '%s' -> %s%s", addr, addrsym, passed ? "passed" : "failed", verbose || !passed ? " " : "\n" );
+    if( verbose || !passed )
+    {
+        if( !result )
+            printf( "(could not get symbol information for address %p)\n", addr );
+        else
+            print_dl_info( &info, "(", ")\n");
+    }
+    return !passed;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _MSC_VER
+#include <windows.h>
+#elif _WIN32
+/* for the following functions force a specific import type regardless of the include file in which they are defined */
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+/* Link by import thunk */
+HMODULE __stdcall GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T __stdcall VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+/* Link by iat */
+__declspec(dllimport) HMODULE __stdcall LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
+#endif
```

Comment:
I have one another question for this part of code.

Cannot it be simplified to just only this?
```c
#ifdef _WIN32
#include <windows.h>
#ifndef _MSC_VER
/* Link by import thunk */
HMODULE __stdcall GetModuleHandleA (LPCSTR lpModuleName);
SIZE_T __stdcall VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
/* Link by iat */
__declspec(dllimport) HMODULE __stdcall LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
#endif
#endif
```
I think that gcc/mingw should be able to handle "overwriting" existing function with just different linkage specified by `__declspec(dllimport)`.

Reply:
There will still be a warning. A better solution would be to 'undefine' the prototypes in question as shown below:
```
#ifdef _WIN32
/* undefine specific prototypes */
#define VirtualQuery _VirtualQuery
#define GetModuleHandleA _GetModuleHandleA
#include <windows.h>
#undef VirtualQuery
#undef GetModuleHandleA
/* Link by import thunk */
HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
#endif
```
That will probably also work with msvc.

BTW: WINAPI is defined as __stdcall

Reply:
> There will still be a warning.

Could you post that warning? Maybe we can find some clean way how to get rid it off.

I do not like those `#define VirtualQuery _VirtualQuery` as it can be very fragile. It needs to be done prior including any other file and also it is fragile...

Reply:
```
~src/dlfcn-win32/tests/test-dladdr.c:115:16: warning: 'GetModuleHandleA' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
                ^~~~~~~~~~~~~~~~
~/src/dlfcn-win32/tests/test-dladdr.c:116:15: warning: 'VirtualQuery' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
```

Reply:
Ok, above warning is just false-positive alarm. I think it is really bad idea to start inventing fragile workaround to just mute false-positive alarm, which may have side effects.

Personally I would ignore such warning and put comment into code explaing "this is really what we want becaue ... and compiler probably generate false-positive warning because it thinks we wanted something else...".

But for muting this warning, something like this should work for gcc:

```c
#ifdef _WIN32
#include <windows.h>
#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wattributes"
/* Link by import thunk */
HMODULE __stdcall GetModuleHandleA (LPCSTR lpModuleName);
SIZE_T __stdcall VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
/* Link by iat */
__declspec(dllimport) HMODULE __stdcall LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
#pragma GCC diagnostic pop
#endif
#endif
```

What do you think about it? Can you test if it works or not?

Reply:
Yes, it works - the compiler does not complain anymore.

Reply:
Can you enable this code also for MSVC (without pragmas) and post what is the msvc error or warning?

Reply:
For MSVC you should be able to mute that warning by:
```c
#pragma warning(push)
#pragma warning(disable: 4273)
/* code with disabled warning C4273 inconsistent dll linkage */
#pragma warning(pop)
```